neuralpp/symbolic/expression.py

```diff
@@ -0,0 +1,151 @@
+from typing import Any, List
+from abc import ABC, abstractmethod
+
+
+class Expression(ABC):
+    @abstractmethod
+    # 'Expression': forward reference for recursive type reference https://peps.python.org/pep-0484/#forward-references
+    def subexpression(self) -> List['Expression']:
+        """
+        Returns a list of subexpression.
+        E.g.,
+        subexpression(f(x,y)) = [f,x,y]
+        subexpression(add(a,1)) = [add,a,1]
+        """
+        pass
+
+    @abstractmethod
+    def set(self, i: int, new_expression: 'Expression') -> 'Expression':
+        """
+        Set i-th subexpression to new_expression. Count from 0.
+        E.g.,
+        f(x,y).set(1,z) = [f,z,y].
+        If it's out of the scope, return error.
+        """
+        pass
+
+    @abstractmethod
+    def replace(self, from_expression: 'Expression', to_expression: 'Expression') -> 'Expression':
+        """
+        Every expression is immutable so replace() returns either self or a new Expression.
+        No in-place modification should be made.
+        If from_expression is not in the expression, returns self.
+        """
+        pass
+
+    def contains(self, target: 'Expression') -> bool:
+        """
+        Checks if `target` is contained in `self`. The check is deep. E.g.,
+        f(x,f(a,b)).contains(a) == True
+        """
+        for sub_expr in self.subexpression():
+            if sub_expr == target or sub_expr.contains(target):
+                return True
+        return False
+
+    @abstractmethod
+    def __eq__(self, other):
+        pass
+
+
+class Variable(Expression):
+    def __init__(self, name: str):
+        self._name = name
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    def subexpression(self) -> List[Expression]:
+        return []
+
+    def replace(self, from_expression: Expression, to_expression: Expression) -> Expression:
+        return self
+
+    def set(self, i: int, new_expression: Expression) -> Expression:
+        raise IndexError("Variable has no subexpression, so you cannot set().")
+
+    def __eq__(self, other):
+        match other:
+            case Variable(name=other_name):
```

Comment:
I think we want to have `BasicConstant("a") != BasicVariable("a")` and `BasicConstant("a") == SymPyConstant("a")`, so we may still want to stick to this pattern. Alternatively, I can define an `base_type` interface in `AtomicExpression` and use that for comparison instead of `type`

Reply:
per discussion with @rodrigodesalvobraz, use the `base_type` design.